Bobjwalker/project recommendations (#1809)
OctopusDeploy/docs: docs/shared-content/octopus-recommendations/project-recommendations.include.md

When components are decoupled from one another, they can have a different deployment schedules, and do not have to be deployed in a specific order.  That will negate the need for an "orchestration project."   

:::hint
In practice, it is rare for us to see the decoupling of all the components in a web application with a front-end, back-end, and database.  It is much more common for pieces of functionality, or backend services, to be decoupled.  
:::

## Use Lifecycles and Channels to reflect your branching strategies

Most branching strategies follow the "main branch should always be ready to deploy to production" rule.  No changes can be made directly to the main branch.  Instead, work must be done in a branch and merged into main.  Your lifecycles and channels should reflect that rule.

For example, in your Octopus instance you have the following environments.

- Dev
- QA
- Staging
- Production

For most branching strategies, we'd recommend two lifecycles in this example, each with two environments.

- Development Lifecycle
    - Dev
    - QA
- Release Lifecycle
    - Staging
    - Production

> The workflow would be:
> 1. Create a branch, commit some changes.
> 2. Build is triggered on branch check-in.  It creates a release in Octopus for the Development lifecycle and pushes to Dev.
> 3. Changes are verified in Dev and are promoted to QA.
> 4. Full test suite is run in QA.
> 5. Bugs or changes are found, repeat the previous steps.
> 6. After a few iterations the change is ready for Production.
> 7. Create a pull request and merge into main.
> 8. Build is triggered on check-in to main.  It creates a release in Octopus for the Release lifecycle and pushes to Staging.
> 9. Automated tests are run in Staging.
> 10. Assuming tests pass, promote to Production.  If tests don't pass then a new branch is created and this process starts all over.

Reviewer: ```suggestion
The workflow would be as follows:
1. Create a branch, commit some changes.
2. Build is triggered on branch check-in.  It creates a release in Octopus for the Development lifecycle and pushes to Dev.
3. Changes are verified in Dev and are promoted to QA.
4. Full test suite is run in QA.
5. Bugs or changes are found; repeat the previous steps.
6. After a few iterations, the change is ready for Production.
7. Create a pull request and merge into main.
8. Build is triggered on check-in to main.  It creates a release in Octopus for the Release lifecycle and pushes to Staging.
9. Automated tests are run in Staging.
10. Assuming tests pass, promote to Production.  If tests don't pass, then a new branch is created, and this process starts all over.
```

> Octopus Deploy provides the capability for dynamic package / docker image selection.  Meaning, you can have a different package per environment.  The intended use case for this is when using a third-party external feed and feed changes between environments.  The external feed provides the capabilities to "promote" packages ready for deployment.
> 
> We do not recommend having a single lifecycle with all environments.  When that happens, we will see customers create a single release, and then change the package or package version when going from QA to Staging.  Such an approach is very hard to audit and track.  
> 
> Changes made on feature or short-lived branches are not ready for Production.  They should be deployed to testing environments for verification and testing, but they should never have the chance to make it to Production.  Merging into main should trigger a fresh build because you could be merging multiple changes from different branches for the first time.  The underlying code has changed and a new build is needed to test and verify.
> 
> For the packages / docker containers built off of branches, append a pre-release tag to the release version.  Leverage channel version rules to only allow packages / docker containers with a pre-release tag for the Development lifecycle.  At the same time, only allow packages / docker containers WITHOUT a pre-release tag for the release lifecycle.
> 
> :::hint
> This section is another reason why we recommend deploying all tightly coupled components stored in the same source control repository within the same project.  Attempting to coordinate different lifecycles and releases across multiple projects can add aditional overhead which runs the risk of something not working correctly.
> :::

Reviewer: ```suggestion
Octopus Deploy provides the capability for dynamic package / docker image selection.  This allows you to have a different package per environment.  The intended use case is when using a third-party external feed and the feed changes between environments.  The external feed provides the capabilities to "promote" packages ready for deployment.

We don't recommend having a single lifecycle with all environments.  When that happens, we have seen customers create a single release and change the package or package version from QA to Staging.  Such an approach is challenging to audit and track.  

Changes made on feature or short-lived branches are not ready for Production.  They should be deployed to testing environments for verification and testing, but they should never have the chance to make it to Production.  Merging into main should trigger a fresh build because you could be merging multiple changes from different branches for the first time.  The underlying code has changed, and a new build is needed to test and verify.

For the packages / docker containers built from branches, append a pre-release tag to the release version.  Leverage channel version rules to only allow packages / docker containers with a pre-release tag for the Development lifecycle.  At the same time, only allow packages / docker containers **without** a pre-release tag for the release lifecycle.

:::hint
This section is another reason we recommend deploying all tightly coupled components stored in the same source control repository within the same project.  Attempting to coordinate different lifecycles and releases across multiple projects can add aditional overhead, which runs the risk of something needing to be fixed.
:::
```

## Automate every component's deployment {#automate-every-components-deployment}

> A typical scenario we see is that application deployments are automated, but the database piece is not.  That's still a manual process which means a DBA has to run the scripts on the night of deployment to production.  After they finish, the automated process can be kicked off.  Because this is manual, there's a good chance one or more of the scripts were not included in the deployments to dev, testing or staging.  Without prior testing, the likelihood of success goes down, and the deployment time goes up.

Reviewer: ```suggestion
A typical scenario we see is that application deployments are automated, but the database deployment is manual.  This means a DBA must run the scripts on the night of deployment to production.  After they finish, the automated process can be kicked off.  Because this is manual, there's a good chance one or more of the scripts were not included in the deployments to dev, testing or staging.  Without prior testing, the likelihood of success decreases and the deployment time increases.
```
